core/src/main/java/com/linecorp/armeria/common/stream/DeferredStreamMessage.java

```diff
@@ -250,6 +250,7 @@ private void doCancel() {
             }
         } else {
             cancelPending = true;
+            abort(CancelledSubscriptionException.get());
```

Comment:
Do we need this?
Isn't it enough to call `upstreamSubscription.cancel();` here?
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/common/stream/DeferredStreamMessage.java#L387

Reply:
`upstreamSubscription` is lazily initialized, so it is null in the `else` block.
The conditional does null check for `upstreamSubscription`.

Reply:
Oh,I got it wrong about the comment. 

`upstreamSubscription.cancel()` can never be called before the backing future completes, as the subscription gets lazily initialized once its backing future completes.

But, we may want to cancel the pending future before it completes once `DeferredStreamMessage` gets cancelled.

Reply:
> can never be called before the backing future completes, as the subscription gets lazily initialized once its backing future completes.

IIRC, that was the intention of the `DeferredStreamMessage` so that `upstream.cancel()` is propagated correctly when the `upstream` is set.
However, aborting the `DeferredStreamMessage` as soon as possible seems a good approach as well. 👍

Reply:
Now that we abort `DeferredStreamMessage` even without its backing future completed, could we remove `cancelPending` flag and its related logic?

Build is successful even after I do so, but still not sure if there's any side effect.

Reply:
I have checked the execution flow and yeah we can remove the flag. 😉